joint_trajectory_controller/include/joint_trajectory_controller/tolerances.hpp

```diff
@@ -106,16 +109,149 @@ SegmentTolerances get_segment_tolerances(Params const & params)
+    // map joint names from goal to active_tolerances
+    auto it = std::find(params.joints.begin(), params.joints.end(), joint);
+    if (it == params.joints.end())
+    {
+      RCLCPP_ERROR(
+        logger, "%s",
+        ("joint " + joint + "specified in goal.path_tolerance does not exist").c_str());
+      return active_tolerances;
+    }
+    auto i = std::distance(params.joints.cbegin(), it);
+    if (joint_tol.position > 0.0)
+    {
+      active_tolerances.state_tolerance[i].position = joint_tol.position;
+    }
+    else if (joint_tol.position == -1.0)
+    {  // -1 means erase
+      active_tolerances.state_tolerance[i].position = 0.0;
+    }
+
+    if (joint_tol.velocity > 0.0)
+    {
+      active_tolerances.state_tolerance[i].velocity = joint_tol.velocity;
+    }
+    else if (joint_tol.velocity == -1.0)
+    {  // -1 means erase
+      active_tolerances.state_tolerance[i].velocity = 0.0;
+    }
+
+    if (joint_tol.acceleration > 0.0)
+    {
+      active_tolerances.state_tolerance[i].acceleration = joint_tol.acceleration;
+    }
+    else if (joint_tol.acceleration == -1.0)
+    {  // -1 means erase
+      active_tolerances.state_tolerance[i].acceleration = 0.0;
+    }
+
+    RCLCPP_DEBUG(
+      logger, "%s %f", (joint + ".state_tolerance.position").c_str(),
+      active_tolerances.state_tolerance[i].position);
+    RCLCPP_DEBUG(
+      logger, "%s %f", (joint + ".state_tolerance.velocity").c_str(),
+      active_tolerances.state_tolerance[i].velocity);
+    RCLCPP_DEBUG(
+      logger, "%s %f", (joint + ".state_tolerance.acceleration").c_str(),
+      active_tolerances.state_tolerance[i].acceleration);
+  }
+  for (auto goal_tol : goal.goal_tolerance)
+  {
+    auto const joint = goal_tol.name;
+    // map joint names from goal to active_tolerances
+    auto it = std::find(params.joints.begin(), params.joints.end(), joint);
+    if (it == params.joints.end())
+    {
+      RCLCPP_ERROR(
+        logger, "%s",
+        ("joint " + joint + "specified in goal.goal_tolerance does not exist").c_str());
+      return active_tolerances;
+    }
+    auto i = std::distance(params.joints.cbegin(), it);
+    if (goal_tol.position > 0.0)
+    {
+      active_tolerances.goal_state_tolerance[i].position = goal_tol.position;
+    }
+    else if (goal_tol.position == -1.0)
+    {  // -1 means erase
+      active_tolerances.goal_state_tolerance[i].position = 0.0;
+    }
+
+    if (goal_tol.velocity > 0.0)
+    {
+      active_tolerances.goal_state_tolerance[i].velocity = goal_tol.velocity;
+    }
+    else if (goal_tol.velocity == -1.0)
+    {  // -1 means erase
+      active_tolerances.goal_state_tolerance[i].velocity = 0.0;
+    }
+
+    if (goal_tol.acceleration > 0.0)
+    {
+      active_tolerances.goal_state_tolerance[i].acceleration = goal_tol.acceleration;
+    }
+    else if (goal_tol.acceleration == -1.0)
+    {  // -1 means erase
+      active_tolerances.goal_state_tolerance[i].acceleration = 0.0;
+    }
```

Comment:
I've a question here, why only -1 is considered as erase tolerance?. I see that if it is positive, we are considering the tolerance. So why not like any value less than zero is considered as no active tolerances?

Reply:
because it is described [here](https://github.com/ros-controls/control_msgs/blob/b4354a10acef592bd0cabf6092f95b843b7df50e/control_msgs/msg/JointTolerance.msg#L7-L10), I thought this might be the same behavior as with ros-controls?

Reply:
It could be from ros_control as you said. 
For me it makes sense to rule out all negative values and count them as disabling the tolerance. 

What's your opinion on this?

Reply:
I don't have a strong opinion on that. @bmagyar?

Reply:
I actually have an opinion on that. In general any random number that's negative should be an *invalid value* but we are making an exception with `-1.0` which is one of an infinite number of negative floating point number to do a specific thing.

Reply:
Do you want me to add an error message if an arbitrary negative value is sent?

Reply:
Yes please

Reply:
Done, and added more tests

```
$ colcon test --packages-select joint_trajectory_controller
Starting >>> joint_trajectory_controller
[Processing: joint_trajectory_controller]                   
[Processing: joint_trajectory_controller]                           
[Processing: joint_trajectory_controller]                             
[Processing: joint_trajectory_controller]                             
[Processing: joint_trajectory_controller]                             
[Processing: joint_trajectory_controller]                             
[Processing: joint_trajectory_controller]                             
[Processing: joint_trajectory_controller]                             
[Processing: joint_trajectory_controller]                             
[Processing: joint_trajectory_controller]                             
Finished <<< joint_trajectory_controller [5min 1s]                  

Summary: 1 package finished [5min 2s]
vscode ➜ /workspaces/ros2_iron_ws (master ✗) $ colcon test-result
Summary: 608 tests, 0 errors, 0 failures, 0 skipped
```